AY2021S1-CS2103-T16-3/tp
# Refactoring with Allocation association class and Bug fix for delete/edit students' allocated rooms

### What this does
This PR refactors and removes the cyclic dependency between `Student` and `Room` using an `Allocation` association class that stores identifiers for `Student` and `Room`, and introduces a bug fix when `Delete` and `Edit` student commands do not update the student's allocated rooms.


Closes #93 and fixes #90.

### How to test
1. Launch the application on an empty `/data` folder.
2. Run `allocate ri/2 si/2`. You should expect to see `Room 11:104 allocated to Bernice Yu.`. Exit the application and view `data/addressbook.json`. You should expect to see 2 entries in `allocations`. This shows that the `allocate` command works correctly.
3. Launch the application again, and run `deallocate si/1`. You should expect to see `Room 11:108 deallocated for Alex Yeoh. Exit the application and view `data/addressbook.json`. You should expect to see 1 entry in `allocations`. This shows that the `deallocate` command works correctly.
4. Launch the application again, and run `reallocate si/2 ri/1`. You should expect to see `Student Bernice Yu has been reallocated to Room 11:108`. Exit the application and view `data/addressbook.json`. You should expect to see that the entry in `allocations` has been changed to `roomNumber: 108`. This shows that the `reallocate` command works correctly.
5. Launch the application again, and click the `Rooms` tab. You should see room `11-108` allocated to `Bernice Yu` of student ID `E02222222`. Run `delete 2`. Notice that it flags `Deallocate student first!`. This shows the bug fix for delete command.
6. Run `edit 2 i/E0123456`, and click the `Rooms` tab. Click the Room Card for room `11-108`. Notice that it updates the student ID. This shows the bug fix for edit command.
7. Exit the application, and smile to yourself at the likelihood of applying to the module you _love_ as a TA.

### Notes
1. To add testing support for `AllocateCommand`, `DeallocateCommand`, `ReallocateCommand`, `Allocation`, `UniqueAllocationList`, among others.
